core/src/main/java/org/aerogear/mobile/core/http/OkHttpResponse.java

```diff
@@ -46,15 +51,31 @@ private synchronized void runCompletionHandler() {
         }
     }
 
+    private synchronized void runErrorHandler() {
+        if (errorHandler != null) {
+            errorHandler.run();
+            errorHandler = null;
+        }
+    }
+
     @Override
     public HttpResponse onComplete(Runnable completionHandler) {
         this.completionHandler = completionHandler;
-        if (response != null) {
+        if (response != null && response.isSuccessful()) {
             runCompletionHandler();
         }
         return this;
     }
 
+    @Override
+    public HttpResponse onError(Runnable errorHandler) {
+        this.errorHandler = errorHandler;
+        if (response != null && !response.isSuccessful()) {
```

Comment:
I'm not sure if we need this kind of check here. IMO we should check only if `errorHandler` is null to avoid NPE

Reply:
This check is required because otherwise it could happen that you attach the errorHandler when the network thread has not yet finished and response is still null. In that case we can not yet check if the response was successful and have to wait for the network thread callback to handle that.